File: Part 2 - Regression/Section 6 - Polynomial Regression/polynomial_regression.py
#Polynimoal Regression

#Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#import dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values
"""X--> should always be matrix('1' considerd as vector and '1:2' considerd as matrix)
   y--> should always be vector"""

#Spliting the dataset into the training set and the test set
"""from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)"""

#Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScalar()
X_train = sc_X.fit_tarnsform(X_train)
X_test = sc_X.transform(X_test)"""

#Fitting linear regression to the dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y)

#Fitting polynomial regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)

#Visualzing the Linear Regression results
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg.predict(X), color = 'blue')
plt.title('Truth or Bluff (Linear Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

#Visualizing the polynomial regression results
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg_2.predict(poly_reg.fit_transform(X)), color = 'blue')
plt.tilte('Truth or Bluff (Ploynomial Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

#predicting a new result with linear regression
lin_reg.predict(6.5)

#predictiong a nre result with polynomial regression
lin_reg_2.predict(poly_reg.fit_transform(6.5))

File: Part 2 - Regression/Section 6 - Polynomial Regression/polynomial_regression.R
#Polynomial Regression

#importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

#Fitting linear regression to dataset
lin_reg = lm(formula = Salary ~ .,
             data = dataset)

#Fitiing ploynomial regression to dataset
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ .,
              data = dataset)

#Visualizing the linear regression results
library(ggplot2)
ggplot()+
  geom_point(aes(x=dataset$Level, y=dataset$Salary),
           colour='red')+
  geom_line(aes(x=dataset$Level, y=predict(lin_reg, newdata = dataset)),
            colour='blue')+
  ggtitle('Truth vs Bluff (Linear regression)')+
  xlab('Level')+
  ylab('Salary')

#Visualizing the polynomial Regression results
# library(ggplot2)
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot()+
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour='red')+
  geom_line(aes(x = x_grid, y=predict(poly_reg, newdata = data.frame(Level = x_grid))),
            colour='blue')+
  ggtitle('Truth vs Bluff (Polynomial regression)')+
  xlab('Level')+
  ylab('Salary')

#predictiong a new result set with linear regression
y_pred = predict(lin_reg, data.frame(Level = 6.5))

#prdicting a new result set with polynomial regression
y_pred = predict(poly_reg, data.frame(Level = 6.5,
                                      Level2 = 6.5^2,
                                      Level3 = 6.5^3,
                                      Level4 = 6.5^4))
